utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,457 @@
+    <> "array" .= array
+    )
+
+data Ubo = Ubo
+  { name :: Text
+  , set :: Int
+  , binding :: Int
+  , array :: Maybe (Vector Int)
+  } deriving (Show, Generic, FromJSON, ToJSON)
+
+data Texture = Texture
+  { type' :: TextureDescriptorType
+  , name :: Text
+  , set :: Int
+  , binding :: Int
+  , array :: Maybe (Vector Int)
+  } deriving (Show, Generic)
+
+instance FromJSON Texture where
+  parseJSON = withObject "textures" $ \v -> Texture
+    <$> v .: "type"
+    <*> v .: "name"
+    <*> v .: "set"
+    <*> v .: "binding"
+    <*> v .:? "array"
+
+instance ToJSON Texture where
+  toJSON (Texture type' name set binding array) = object
+    [ "type" .= type'
+    , "name" .= name
+    , "set" .= set
+    , "binding" .= binding
+    , "array" .= array
+    ]
+  toEncoding (Texture type' name set binding array) = pairs
+    (  "type" .= type'
+    <> "name" .= name
+    <> "set" .= set
+    <> "binding" .= binding
+    <> "array" .= array
+    )
+
+data Reflection = Reflection
+  { entryPoints :: Vector EntryPoint
+  , inputs :: Maybe (Vector Input)
+  , textures :: Maybe (Vector Texture)
+  , ubos :: Maybe (Vector Ubo)
+  } deriving (Show, Generic, FromJSON, ToJSON)
+
+data ConvertException = ConvertException Text Text
+  deriving (Show)
+
+instance Exception ConvertException
+
+class Convert a where
+  eitherFrom :: Text -> Either ConvertException a
+  from :: Text -> a
+  from x = case eitherFrom x of
+    Left ex -> throw ex
+    Right res -> res
+  to :: a -> Text
+
+data ShaderStage
+  = Vert
+  | Frag
+  | Comp
+  | Tesc
+  | Tese
+  | Geom
+  | Rgen
+  | Rint
+  | Rahit
+  | Rchit
+  | Rmiss
+  | Rcall
+  | Task
+  | Mesh
+  deriving (Eq, Show)
+
+instance Convert ShaderStage where
+  eitherFrom = \case
+    "vert" -> Right Vert
+    "frag" -> Right Frag
+    "comp" -> Right Comp
+    "tesc" -> Right Tesc
+    "tese" -> Right Tese
+    "geom" -> Right Geom
+    "rgen" -> Right Rgen
+    "rint" -> Right Rint
+    "rahit" -> Right Rahit
+    "rchit" -> Right Rchit
+    "rmiss" -> Right Rmiss
+    "rcall" -> Right Rcall
+    "task" -> Right Task
+    "mesh" -> Right Mesh
+    unsupport -> Left $ ConvertException unsupport $ "ShaderStage not support '" <> unsupport <> "'"
+  to = \case
+    Vert -> "vert"
+    Frag -> "frag"
+    Comp -> "comp"
+    Tesc -> "tesc"
+    Tese -> "tese"
+    Geom -> "geom"
+    Rgen -> "rgen"
+    Rint -> "rint"
+    Rahit -> "rahit"
+    Rchit -> "rchit"
+    Rmiss -> "rmiss"
+    Rcall -> "rcall"
+    Task -> "task"
+    Mesh -> "mesh"
+
+instance FromJSON ShaderStage where
+  parseJSON value@(String x) = case eitherFrom @ShaderStage x of
+    Left (ConvertException _e err) -> prependFailure (T.unpack $ err <> ", ") . unexpected $ value
+    Right _stage -> withText "mode" (pure . from) value
+  parseJSON value = withText "mode" (pure . from) value
+
+instance ToJSON ShaderStage where
+  toJSON = String . to
+  toEncoding = toEncoding . to
+
+convertStage :: ShaderStage -> ShaderStageFlagBits
```

Comment:
The output type is quite suggestive already.

```haskell
shaderStageFlagBits :: ShaderStage -> ShaderStageFlagBits
```

This will avoid raising questions like "converting stage into what?.."

Reply:
`convertStage` is renamed to `shaderStageFlagBits`.

The following functions with `convert` prefix is not renamed.
1. `convertTextureDescriptorType :: TextureDescriptorType -> DescriptorType`, this renamed to `descriptorType` will cause a "Ambiguous occurrence" error
2. `convertVertexAttributeType :: VertexAttributeType -> (Word32, Format)`, not sure which new name to use, and `format` will also cause a "Ambiguous occurrence" error